keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,520 @@
+        // +optional
+	PodSelector *metav1.LabelSelector
+
+        // ContainerSelector is a selector which must be true for the hook to run on the container
+        // Hook controller will find all containers in the selected pods based on ContainerSelector,
+        // if specified.
+        // The ContainerSeletor maps pod label to container name(s) so that the controller only select
+	// the containers that are listed in the map. Otherwise, all containers of the pods that match
+	// the pod label selector will be selected.
+        // +optional
+        ContainerSelector map[string]string
+ }
+```
+
+Here is the definition of the ExecutionHookStatus. This represents the current state of a hook for all selected containers in the selected pods.
+
+```
+// ExecutionHookStatus represents the current state of a hook
+type ExecutionHookStatus struct {
+        // This is a list of ContainerExecutionHookStatus, with each status representing information
+        // about how hook is executed in a container, including pod name, container name, ActionTimestamp,
+        // ActionSucceed, etc.
+        // +optional
+        ContainerExecutionHookStatuses []ContainerExecutionHookStatus
+
+        // Action Summary status
+        // Default is nil
+        // +optional
+        ActionSucceed *bool
+}
+
+// ContainerExecutionHookStatus represents the current state of a hook for a specific container in a pod
+type ContainerExecutionHookStatus struct {
+        // This field is required
+        PodName string
+
+        // This field is required
+        ContainerName string
+
+        ActionStatus *ExecutionHookActionStatus
+ }
+
+type ExecutionHookActionStatus struct {
+        // If not set, it is nil, indicating Action has not started
+        // If set, it means Action has started at the specified time
+	// +optional
+        ActionTimestamp *int64
+
+        // If not set, it is nil, indicating Action is not complete on the specified container in the pod
+        // +optional
+        ActionSucceed *bool
+
+        // The last error encountered when running the action
+	// +optional
+        Error *HookError
+}
+
+type HookError storage.VolumeError
+```
+
+In the ExecutionHookStatus object, there is a list of ContainerExecutionHookStatus for all selected containers in the pods, each ContainerExecutionHookStatus represents the state of the hook on a specific container.
+
+The ActionSucceed field is a summary status field to indicate the final status of the Action in the hook after the commands are run on all selected containers. If not set, this field is nil, meaning the hook controller still needs to wait for the Action commands to finish running on all containers. If set, either true or false, it means the Action commands have finished running (or timeout) on all containers.
+
+
+Here is the definition for PodAction:
+
+```
+// PodAction describes action commands to run on pods/containers based
```

Comment:
Are we expecting users to pre-define `PodAction` objects to be used later, or snapshot controllers (application snapshot controller or volume snapshot controller) to create them at runtime? If it's the latter, making `PodAction` a separate CRD makes less sense because the controllers have to create two objects: creating a `PodAction` object first and then an `ExecutionHook` object.

Reply:
what is the difference for users? I am thinking the userwill pre-define PodAction

Reply:
I thought the idea is for user to create a PodAction that can be re-used later.

Reply:
OK. Guess we should make this clear in the doc.

Reply:
I'll add comments.